src/main/java/com/uid2/admin/vertx/service/EncryptionKeyService.java

```diff
@@ -157,11 +170,14 @@ private void handleKeyList(RoutingContext rc) {
 
     private void handleRotateMasterKey(RoutingContext rc) {
         try {
-            final RotationResult result = rotateKeys(rc, masterKeyActivatesIn, masterKeyExpiresAfter,
-                    s -> s == Const.Data.MasterKeySiteId || s == Const.Data.RefreshKeySiteId);
+            final RotationResult masterKeyResult = rotateKeys(rc, masterKeyActivatesIn, masterKeyExpiresAfter, masterKeyRotationCutoffTime,
+                s -> s == Const.Data.MasterKeySiteId);
+            final RotationResult refreshKeyResult = rotateKeys(rc, masterKeyActivatesIn, masterKeyExpiresAfter, refreshKeyRotationCutOffTime,
+                    s -> s == Const.Data.RefreshKeySiteId);
```

Comment:
so each rotateKeys call will indirectly call s3 upload once so essentially if you call rotateKeys twice you are doing 2 updates/uploads for encryption key files. it doesn't really matter as it's called within a single writeLock so there won't be inconsistency issue i think (e.g. master is updated but not refresh keys yet and an operator is trying to download them all). but it's an inefficiency. if you are in a hurry i suggest just create a jira ticket for future improvement.

or if you wanna fix it maybe you can create a Predicate CutoffTimeSelector (similar to the Predicate<Integer> siteSelector) that will take in siteId and return cutoff time accordingly.

Reply:
there will be a race condition over max_id (metadata in memory is updated by downloading the new metadata file I suppose, so continuous writing without fetching will create keys with same id and conflict)

Reply:
best back to having 1 rotateKeys call then